Dockerfile

```diff
@@ -21,19 +21,19 @@ ENV PATH=$PATH:/opt/mssql-tools/bin
 RUN \
   mkdir /app && \
   mkdir /workspace
-COPY . /app
-
-# We have to set this because requirements.txt contains relative path
-# references
-WORKDIR /app
+COPY ./requirements.txt /app/requirements.txt
+RUN \
+  python -m pip install --upgrade 'pip>=21,<22' && \
+  python -m pip install --requirement /app/requirements.txt
 
+# Install the application
+COPY . /app
```

Comment:
or add more stuff to .dockerignore?

Reply:
That would be good, but I don't immediately know what things we actually need and I don't want to make this into yet another diversion. The entire repo is 2.4M so it's not going to be an enormous win in any case, though I agree it would be nice. The main thing would be to get rid of the 1.3M `sqljdbc_9.2` directory, but I fear that might be needed at runtime by the pyspark lib.

Reply:
Okay, no problem. I agree you shouldn't get diverted.

Reply:
So, docker will rebuild this (and subsequent layers ofc) if any file in `.` has changed, modulo `.dockerignore` exclusions. So copying less/exlcuding more is faster not primarily because of the size, but avoiding a cache miss.

Reply:
This layer is _always_ going to change every time anyway

Reply:
Right, but right now it would rebuild if you change anything in `.github/`, or `docker-compose.yaml`, or any `.md` file, none of which require updating the docker image.

Reply:
It will only rebuild if you bump the release version with a `fix:` or `feature:` commit, which you wouldn't do if you'd only changed `docker-compose.yaml`. What am I missing?

Reply:
I mean rebuild locally in dev

Reply:
Every layer will always rebuild in CI, as there's no local cache

Reply:
Ah, right! Sorry, was totally misunderstanding :)

Yes, I see what you mean. It hasn't made much difference to me because I use the `run.sh` script which mounts the code in and only rebuild occasionally. But yeah, that would be another advantage of tidying this up.

Reply:
Does run use `docker-compose run`? I thought that would rebuild automatically?

Reply:
It does use `docker-compose run` but there's no automatic rebuilding:
> Services are built once and then tagged, by default as `project_service` ...
> If you change a service’s Dockerfile or the contents of its build directory, run `docker-compose build` to rebuild it.

https://docs.docker.com/compose/reference/build/